### PR TITLE
feat(backend): read communities data from database

### PR DESCRIPTION
**Pull request by clauspeterhuebner**
### Issues

- None

### Todo

- [X] None

### CODE REVIEW

File: backend/src/graphql/resolver/CommunityResolver.test.ts [hunk 1]
```diff
@@ -4,118 +4,234 @@
...
 /* eslint-disable @typescript-eslint/explicit-module-boundary-types */
 
-import { createTestClient } from 'apollo-server-testing'
-import createServer from '@/server/createServer'
-import CONFIG from '@/config'
+import { Community as DbCommunity } from '@entity/Community'
+import { testEnvironment } from '@test/helpers'
 
-jest.mock('@/config')
+// jest.mock('@/config')
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolver/CommunityResolver.test.ts:10]: remove

File: backend/src/graphql/resolver/CommunityResolver.test.ts [hunk 2]
```diff
@@ -4,118 +4,234 @@
...
+  await DbCommunity.clear()
 })
 
 afterAll(async () => {
   await con.close()
 })
 
 describe('CommunityResolver', () => {
-  const getCommunityInfoQuery = `
+  const getCommunities = `
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolver/CommunityResolver.test.ts:30]: please outsource to `seeds/graphql/queries.ts`

File: backend/src/graphql/resolver/CommunityResolver.ts [hunk 3]
```diff
@@ -1,58 +1,21 @@
...
-        id: 3,
-        name: 'Gradido-Akademie',
-        description: 'Freies Institut für Wirtschaftsbionik.',
-        url: 'https://gradido.net',
-        registerUrl: 'https://gdd1.gradido.com/register-community',
-      }),
-    ]
+  async getCommunities(): Promise<Community[]> {
+    const comList: Community[] = []
+    const dbCommunities: DbCommunity[] = await DbCommunity.find({ order: { id: 'ASC' } })
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolver/CommunityResolver.ts:14]: sort by pubkey + endpoint

File: backend/src/graphql/resolver/CommunityResolver.ts [hunk 4]
```diff
@@ -1,58 +1,21 @@
...
 
 import { Community } from '@model/Community'
+import { Community as DbCommunity } from '@entity/Community'
 
 import { RIGHTS } from '@/auth/RIGHTS'
-import CONFIG from '@/config'
 
 @Resolver()
 export class CommunityResolver {
-  @Authorized([RIGHTS.GET_COMMUNITY_INFO])
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolver/CommunityResolver.ts:10]: remove `RIGHTS.GET_COMMUNITY_INFO`

File: backend/src/graphql/resolver/CommunityResolver.ts [hunk 5]
```diff
@@ -1,58 +1,21 @@
...
-  @Query(() => Community)
-  getCommunityInfo(): Community {
-    return new Community({
-      name: CONFIG.COMMUNITY_NAME,
-      description: CONFIG.COMMUNITY_DESCRIPTION,
-      url: CONFIG.COMMUNITY_URL,
-      registerUrl: CONFIG.COMMUNITY_REGISTER_URL,
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolver/CommunityResolver.ts:17]: Is this still in the config?

File: backend/src/graphql/resolver/CommunityResolver.ts [hunk 6]
```diff
@@ -1,58 +1,21 @@
...
-        name: 'Gradido-Akademie',
-        description: 'Freies Institut für Wirtschaftsbionik.',
-        url: 'https://gradido.net',
-        registerUrl: 'https://gdd1.gradido.com/register-community',
-      }),
-    ]
+  async getCommunities(): Promise<Community[]> {
+    const comList: Community[] = []
+    const dbCommunities: DbCommunity[] = await DbCommunity.find({ order: { id: 'ASC' } })
+    dbCommunities.forEach(async function (dbCom) {
```
Comment (ulfgebhardt) [inline backend/src/graphql/resolver/CommunityResolver.ts:15]: use map

Comment (ulfgebhardt) [reply backend/src/graphql/resolver/CommunityResolver.ts:15]: example
```
return {
      links: links.map((link: DbContributionLink) => new ContributionLink(link)),
      count,
    }
```

File: backend/src/graphql/resolver/CommunityResolver.test.ts (hunk 1 above)
Comment (clauspeterhuebner) [reply backend/src/graphql/resolver/CommunityResolver.test.ts:10]: done

File: backend/src/graphql/resolver/CommunityResolver.test.ts (hunk 2 above)
Comment (clauspeterhuebner) [reply backend/src/graphql/resolver/CommunityResolver.test.ts:30]: done

File: backend/src/graphql/resolver/CommunityResolver.ts (hunk 3 above)
Comment (clauspeterhuebner) [reply backend/src/graphql/resolver/CommunityResolver.ts:14]: done

File: backend/src/graphql/resolver/CommunityResolver.ts (hunk 4 above)
Comment (clauspeterhuebner) [reply backend/src/graphql/resolver/CommunityResolver.ts:10]: done

File: backend/src/graphql/resolver/CommunityResolver.ts (hunk 5 above)
Comment (clauspeterhuebner) [reply backend/src/graphql/resolver/CommunityResolver.ts:17]: Yes, some of these config properties are still in use and with the next steps of federation they will be used more than now

File: backend/src/graphql/resolver/CommunityResolver.ts (hunk 6 above)
Comment (clauspeterhuebner) [reply backend/src/graphql/resolver/CommunityResolver.ts:15]: done

File: backend/src/graphql/resolver/CommunityResolver.test.ts [hunk 7]
```diff
@@ -1,121 +1,221 @@
...
+let testEnv: any
 
 beforeAll(async () => {
-  const server = await createServer({})
-  con = server.con
-  query = createTestClient(server.apollo).query
+  testEnv = await testEnvironment()
+  query = testEnv.query
+  con = testEnv.con
+  await DbCommunity.clear()
```
Comment (Elweyn) [inline backend/src/graphql/resolver/CommunityResolver.test.ts:22]: Why clear the DbCommunity beforeAll() and not afterAll().
I would build up the Data in DbCommunity beforeAll and reset/clear DbCommunity afterAll.

Comment (Elweyn) [reply backend/src/graphql/resolver/CommunityResolver.test.ts:22]: Or to be sure doing both

File: backend/src/graphql/resolver/CommunityResolver.ts [hunk 8]
```diff
@@ -1,58 +1,18 @@
...
-        id: 3,
-        name: 'Gradido-Akademie',
-        description: 'Freies Institut für Wirtschaftsbionik.',
-        url: 'https://gradido.net',
-        registerUrl: 'https://gdd1.gradido.com/register-community',
-      }),
-    ]
+  async getCommunities(): Promise<Community[]> {
+    const dbCommunities: DbCommunity[] = await DbCommunity.find({
+      order: { foreign: 'ASC', publicKey: 'ASC', apiVersion: 'ASC' },
```
Comment (Elweyn) [inline backend/src/graphql/resolver/CommunityResolver.ts:14]: Do we want to export both foreign and home community?

File: database/entity/0060-update_communities_table/Community.ts [hunk 9]
```diff
@@ -25,13 +25,13 @@ export class Community extends BaseEntity {
   endPoint: string
 
   @Column({ name: 'last_announced_at', type: 'datetime', nullable: true })
-  lastAnnouncedAt: Date
+  lastAnnouncedAt: Date | null
```
Comment (Elweyn) [inline database/entity/0060-update_communities_table/Community.ts:28]: Not certain if ``| null`` could be replaced with 
```suggestion
  lastAnnouncedAt?: Date
```

File: database/entity/0060-update_communities_table/Community.ts [hunk 10]
```diff
@@ -25,13 +25,13 @@ export class Community extends BaseEntity {
...
 
   @Column({ name: 'verified_at', type: 'datetime', nullable: true })
-  verifiedAt: Date
+  verifiedAt: Date | null
```
Comment (Elweyn) [inline database/entity/0060-update_communities_table/Community.ts:31]: Not certain if ``| null`` could be replaced with 
```suggestion
  verifiedAt?: Date
```

File: database/entity/0060-update_communities_table/Community.ts [hunk 11]
```diff
@@ -25,13 +25,13 @@ export class Community extends BaseEntity {
...
 
   @Column({ name: 'last_error_at', type: 'datetime', nullable: true })
-  lastErrorAt: Date
+  lastErrorAt: Date | null
```
Comment (Elweyn) [inline database/entity/0060-update_communities_table/Community.ts:34]: Not certain if ``| null`` could be replaced with 
```suggestion
  lastErrorAt?: Date
```

**Review by Elweyn — COMMENTED** [review]
Except for some small questions and some possible upgrades, it's a great PR!

File: backend/src/graphql/resolver/CommunityResolver.test.ts (hunk 7 above)
Comment (clauspeterhuebner) [reply backend/src/graphql/resolver/CommunityResolver.test.ts:22]: all test cases depend on each other, so only a beforeAll is important to have a clear state of database.
not deleting afterAll was to check the correct values in DB after the tests.

File: backend/src/graphql/resolver/CommunityResolver.ts (hunk 8 above)
Comment (clauspeterhuebner) [reply backend/src/graphql/resolver/CommunityResolver.ts:14]: yes, as long as we have only this community table of federation dynamically filled by the dht-node we show the whole content of this table.
As soon as we will have the second community-Table, which will be different from the community-federation-Table (the migration will rename the current community-table to community-federation), where the permenant community-data will be stored like the community-name/ -uuid/ -etc. there will be a change for the community-List presented in the UI

File: database/entity/0060-update_communities_table/Community.ts (hunk 9 above)
Comment (clauspeterhuebner) [reply database/entity/0060-update_communities_table/Community.ts:28]: Perhaps, i'am not sure too.
It is copied from other entities where the date is optional too

File: database/entity/0060-update_communities_table/Community.ts (hunk 10 above)
Comment (clauspeterhuebner) [reply database/entity/0060-update_communities_table/Community.ts:31]: see above

File: database/entity/0060-update_communities_table/Community.ts (hunk 11 above)
Comment (clauspeterhuebner) [reply database/entity/0060-update_communities_table/Community.ts:34]: see above

**Review by ulfgebhardt — APPROVED** [review]

**Pull request merged**